pkg/crc/oc/oc_cache.go

```diff
@@ -0,0 +1,170 @@
+	"github.com/code-ready/crc/pkg/extract"
+	"github.com/pkg/errors"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+const (
+	OC_CACHE_DIR = "oc"
+	TAR          = "tar.gz"
+	ZIP          = "zip"
+)
+
+// Oc is a struct with methods designed for dealing with the oc binary
+type OcCached struct{}
+
+func (oc *OcCached) EnsureIsCached() error {
+	if !oc.IsCached() {
+		err := oc.cacheOc()
+		if err != nil {
+			return err
+		}
+
+	}
+	return nil
+}
+
+func (oc *OcCached) GetCacheFilepath() string {
+	return filepath.Join(constants.MachineCacheDir, OC_CACHE_DIR)
+}
+
+func (oc *OcCached) IsCached() bool {
+	if _, err := os.Stat(filepath.Join(oc.GetCacheFilepath(), constants.OcBinaryName)); os.IsNotExist(err) {
+		return false
+	}
+	return true
+}
+
+// cacheOc downloads and caches the oc binary into the minishift directory
+func (oc *OcCached) cacheOc() error {
+	if !oc.IsCached() {
+		// Create tmp dir to download the oc tarball
+		tmpDir, err := ioutil.TempDir("", "crc")
+		if err != nil {
+			return err
+		}
+		assetTmpFile, err := download.Download(constants.DefaultOcURL, tmpDir)
+		if err != nil {
+			return err
+		}
+
+		// Extract the tarball and put it the cache directory.
+		binaryPath := ""
+		switch {
+		case strings.HasSuffix(assetTmpFile, TAR):
+			// unzip
+			tarFile := assetTmpFile[:len(assetTmpFile)-3]
+			err = extract.Ungzip(assetTmpFile, tarFile)
+			if err != nil {
+				return errors.Wrapf(err, "Cannot ungzip '%s'", assetTmpFile)
+			}
+
+			// untar
+			err = extract.Untar(tarFile, tmpDir)
+			if err != nil {
+				return errors.Wrapf(err, "Cannot untar '%s'", tarFile)
+			}
+
+			content, err := listDirExcluding(tmpDir, ".*.tar.*")
+			if err != nil {
+				return errors.Wrapf(err, "Cannot list content of '%s'", tmpDir)
+			}
+			if len(content) > 1 {
+				return errors.New(fmt.Sprintf("Unexpected number of files in tmp directory: %s", content))
+			}
+
+			binaryPath = tmpDir
+		case strings.HasSuffix(assetTmpFile, ZIP):
+			contentDir := assetTmpFile[:len(assetTmpFile)-4]
+			err = extract.Unzip(assetTmpFile, contentDir)
+			if err != nil {
+				return errors.Wrapf(err, "Cannot unzip '%s'", assetTmpFile)
+			}
+			binaryPath = contentDir
+		}
+
+		binaryName := constants.OcBinaryName
+		binaryPath = filepath.Join(binaryPath, binaryName)
+
+		// Copy the requested asset into its final destination
+		outputPath := fmt.Sprintf("%s/%s", constants.MachineCacheDir, OC_CACHE_DIR)
+		err = os.MkdirAll(outputPath, 0755)
+		if err != nil && !os.IsExist(err) {
+			return errors.Wrap(err, "Cannot create the target directory.")
+		}
+
+		finalBinaryPath := filepath.Join(outputPath, binaryName)
+		copy(binaryPath, finalBinaryPath)
+		if err != nil {
+			return err
+		}
+
+		err = os.Chmod(finalBinaryPath, 0777)
+		if err != nil {
+			return errors.Wrapf(err, "Cannot make '%s' executable", finalBinaryPath)
+		}
+
+		return nil
+	}
+	return nil
+}
+
+func copy(src, dest string) error {
```

Comment:
Why is caching not done straight into `/home/user/.crc/cache/oc` but it's split into two steps instead? (more of a question rather than a review comment...)

Reply:
@jsliacan first step is to download in the temp location so in case it is no complete then we can start over and also there we extract the payload, second part is mostly just copy the extract binary to location.

Reply:
this is a common pattern. alternatively it could have used a suffix wlike `.part`. we need this to prevent incomplete downloads (at the moment we do support resuming downloads).